ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,246 @@ type clientHandshakeState struct {
 	session         *ClientSessionState
 }
 
+type CacheKeyGenerator interface {
+	Key(net.Addr) string
+}
+
+type ClientHelloConfiguration struct {
+	//Version in the handshake header
+	HandshakeVersion uint16
+
+	//if len == 32, it will specify the client random
+	//Otherwise, the field will be random
+	ClientRandom []byte
+
+	//if RandomSessionID > 0, will overwrite SessionID w/ that many
+	//random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	//These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	//Optional, both must be non-nil, or neither.
+	//Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+func (c *ClientHelloConfiguration) ValidateExtensions() error {
+	for _, ext := range c.Extensions {
+		switch ext.(type) {
+		case PointFormatExtension:
+			for _, format := range ext.(PointFormatExtension).Formats {
+				if format != pointFormatUncompressed {
+					return errors.New(fmt.Sprintf("Unsupported EC Point Format %d", format))
+				}
+			}
+		case SignatureAlgorithmExtension:
+			for _, algs := range ext.(SignatureAlgorithmExtension).getStruct() {
+				found := false
+				for _, supported := range supportedSKXSignatureAlgorithms {
+					if algs.hash == supported.hash && algs.signature == supported.signature {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return errors.New(fmt.Sprintf("Unsupported Hash and Signature Algorithm (%d, %d)", algs.hash, algs.signature))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ClientHelloConfiguration) ModifyConfig(config *Config) *Config {
+	config.NextProtos = []string{}
+	config.CipherSuites = c.CipherSuites
+	config.MaxVersion = c.HandshakeVersion
+	config.ClientRandom = c.ClientRandom
+	config.CurvePreferences = []CurveID{}
+	config.HeartbeatEnabled = false
+	config.ExtendedRandom = false
+	config.ForceSessionTicketExt = false
+	config.ExtendedMasterSecret = false
+	config.SignedCertificateTimestampExt = false
```

Comment:
This is a reasonable original worry. I'm a bigger fan of the dual codepaths. The failure case here is that the config thinks something is supported, but doesn't tell the server. That likely won't cause issues because a server just won't ack the extension in the ServerHello. And if it does, it is a config-change away from being fixed. This code is just trying to handle dumb config inconsistencies gracefully.

Alternatively, shifting the config format significantly sounds like a major pain to every consumer, and rife with potential major errors. 

2 > 1 IMO